phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+                # can short circuit this child under slightly different conditions
+                pass
+
+            else:
+                if child1.level < level or not (child1.isTerminal and child1.allele == "-"):
+                    newChild1 = genomeNode(level=level)
+                    newChild1.isTerminal = child1.isTerminal
+                    newChild1.rate = child1.rate
+
+                    if child1.isTerminal:
+                        newChild1.refNode = child1.refNode
+                        newChild1.allele = child1.allele
+
+                    else:
+                        newChild1.belowNodes = list(child1.belowNodes)
+       
+
+                deleted_leaves_right = self.deleteNodes(rand - rate0, newChild1, remaining_deletions - deleted_leaves_left, level)
+                newRate1 = newChild1.rate
+
+            node.rate = newRate0 + newRate1
+            node.belowNodes = [newChild0, newChild1]
+
+            return deleted_leaves_left + deleted_leaves_right
+
+    def buildTree(self, node, insert_list, pos_left, pos_right):
+
+        if pos_left == pos_right:
+            # then we are at a leaf, we need to build a new leaf node from scratch
+            node.isTerminal = True
+            node.refNode = node
+            if self.codon:
+                node.allele = self.codonAlleles[insert_list[0]]
+            else:
+                
+                node.allele = self.alleles[insert_list[0]]
+                node.rates = self.mutMatrixRepeats[-1]
+
+            node.genomePos = node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            node.insertionRate = self.insertionRate[-1]
+            node.deletionRate = self.deletionRate[-1]
+            node.deletionLength = self.deletionLength[-1]
+            node.insertionLength = self.insertionLength[-1]
+
+            node.rate = -1 * node.rates[node.allele][node.allele]
+            
+
+
+            return node
+
+        else:
+            # cut the list in half and build a tree for each half
+            middle = pos_left + int((pos_right - pos_left) / 2)
+            half_length = int((pos_right - pos_left)/2)
+
+            child0 = self.buildTree(genomeNode(level=node.level), insert_list[:half_length+1], pos_left, middle)
+            child1 = self.buildTree(genomeNode(level=node.level), insert_list[half_length+1:], middle+1, pos_right)
+            node.belowNodes = [child0, child1]
+            node.rate = child0.rate + child1.rate
+            node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            return node
+
+
+    def sampleInsertion(self, rate, node, rand, level):
+
+        # sample a geometric random variable
+        insertion_length = 1 + int(np.log(rand)/np.log(rate))
+        
+        # TODO generate a nucleotide sequence using the correct model
+        insertion = np.random.choice(self.allelesList, insertion_length, self.insertionFrequencies)
+
```

Comment:
Here maybe it would make sense to create a vector similar to self.mutMatrixRepeats, with one entry for each newly created terminal insertion node.

Reply:
Ok - can we talk about this at 2 also? 
I think there is potentially also a similar problem to do with genomePos - for each insertion and terminal node the have node.genomePos = (index, position), where the index is which number insert it is (e.g. (5, 0) means the 1st nucleotide on insert number 5. This is a bit annoying to deal with because it means that all the terminal nodes that refer to a spot on the reference have their genomePos as an int whereas the insertions have a tuple genomePos - so when it comes to referencing things like mutMatrixRepeats[node.genomePos], you need to check that node.genomePos is actually an integer. A solution might be to make all of these things 2 dimensional... do you think this will work?